paddle/fluid/framework/details/all_reduce_op_handle.cc

```diff
@@ -107,14 +107,25 @@ void AllReduceOpHandle::RunImpl() {
           PADDLE_ENFORCE(platform::dynload::ncclAllReduce(
               buffer, buffer, numel, static_cast<ncclDataType_t>(dtype),
               ncclSum, comm, stream));
+          if (!nccl_ctxs_->need_group_call_) cudaStreamSynchronize(stream);
         });
       }
+
       this->RunAndRecordEvent([&] {
-        platform::NCCLGroupGuard guard;
-        for (auto &call : all_reduce_calls) {
-          call();
+        // TODO(Yancey1989): need allreduce operator to avoid this flag
+        if (nccl_ctxs_->need_group_call_) {
+          platform::NCCLGroupGuard guard;
+          for (auto &call : all_reduce_calls) {
+            call();
+          }
+        } else {
+          // only used in executor_type == ParallalGraph, one thread one GPU
+          // TODO(Yancey1989): use allreduce operator to avoid this tricky.
+          PADDLE_ENFORCE(all_reduce_calls.size() == 1UL);
+          all_reduce_calls[0]();
```

Comment:
why is this different?

Reply:
It will hang when using group call of nccl operators, and the example codes of NCCL wouldn't use the gropu call: https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/examples.html#example-1-one-device-per-process-or-thread .